### DISCUSSION TITLE
Installation on CentOS 8 - node v10.21.0 - uWebSockets v18.11.0

### DISCUSSION THREAD

**Discussion by ghost** [uNetworking/uWebSockets.js#409] [Q&A] [answered]
What libraries are needed to compile? I keep running into an issue with the npm and the compile.

**Comment by hst-m** [comment]
I don't know on CentOS which is a pain to work with in general but on Ubuntu 20.04 you only need `sudo apt install -y clang build-essential` then do `git clone --recurse-submodules https://github.com/uNetworking/uWebSockets.js && cd uWebSockets.js && make` or skip all that and do the faster and easier `npm install uNetworking/uWebSockets.js#v18.11.0`

Update: he thought he needed to compile from source to install but all he needed was the npm install and then do `require('uWebSockets.js')`

**Reply by ghost** [reply]
I have run npm and it cannot find the uws.js or I get it's not compatible with my version of node.

**Reply by ghost** [reply]
![image](https://user-images.githubusercontent.com/47580398/102369412-22100500-3f79-11eb-8808-a756bdd9cec2.png)

**Reply by ghost** [reply]
![image](https://user-images.githubusercontent.com/47580398/102369525-47047800-3f79-11eb-8281-1563fb059781.png)

**Reply by hst-m** [reply]
if you installed with the npm method which looks like you did because its in your node_modules folder you only need to do `require('uWebSockets.js')`, looks like you are trying to do `require('/path/to/uws.js')` which is only needed if you did not use npm but then you need to point it to that uws.js file correctly which looks like you did not do

**Comment by ghost** [comment, marked as answer]
The front page example uses require('uWebSockets.js') and all examples use that. Are the binaries not compatible with CentOS 8?

**Reply by hst-m** [reply]
all the /examples use `require('../dist/uws.js')` which could be source of confusion

**Reply by ghost** [reply]
Alexhultman, thank you for the response; I got everything working and everything works on CentOS. It took me a moment to get things set up correctly and how to instantiate it. After hst-m explained the differences, I was up and going. I am trying to learn this and it's a little lacking in information, and I am a semi-novice programmer in javascript. I understand c# and c++. I may have some questions. I am going to try to do WebRTC through this for the app I am attempting to build.

**Reply by ghost** [reply]
This is what I found to learn the libarary

https://edisonchee.com/writing/intro-to-%C2%B5websockets.js/

**Reply by hst-m** [reply]
@precision1980  there was another guy using that same "tutorial" and it caused him to have error because the example repo was bad https://github.com/uNetworking/uWebSockets.js/discussions/385 People must be googling "uwebsocket.js tutorial" and getting that page or something, dont use these random old tutorials from other people just post questions here

**Reply by ghost** [reply]
Ok, thanks, @hst-m !

**Reply by ghost** [reply]
@hst-m, do you think it is possible to do WebRTC through here?

**Reply by hst-m** [reply]
WebRTC includes SIP, XMPP, and MQTT protocols which are done or can be done over websockets so yes you can use uWebSockets.js for those things

**Reply by ghost** [reply]
Ah. You're right the examples directly use uws.js

**Reply by ghost** [reply]
@alexhultman, @hst-m: are there any good client-server examples? I have a web server, web01, then I have node.js server, uWebSocket01, when I do a client ws from web01 to uWebSocket01 it just fails. However, if run the get and publish as web, no issues.

**Reply by hst-m** [reply]
open new discussion topic and be more specific, not sure what you are saying. you are using the [ws](https://www.npmjs.com/package/ws) client library?  nothing you just said makes sense to me

**Reply by ghost** [reply]
@hst-m - I figured it out:

server.js
```
const $uWS = require("uWebSockets.js");
const port = 8080;

let SOCKETS = [];

const $app = $uWS.App()
    .ws('/*', {
        // config
        compression: 0,
        maxPayloadLength: 16 * 1024 * 1024,
        idleTimeout: 60,

        open: (ws, req) => {
            // this handler is called when a client opens a ws connection with the server
            console.log(`client open`);
        },

        message: ws => {
            // called when a client sends a message
            console.log(`client sent a message`);

        },

        close: (ws, code, message) => {
            // called when a ws connection is closed
            console.log(`connection is closed`);

        }
    }).listen(port, token => {
        token ?
            console.log(`Listening to port ${port}`) :
            console.log(`Failed to listen to port ${port}`);
    });
```

client.js
```
const ws = new WebSocket("ws://10.0.4.7:8080");
const $x = document.getElementById();

ws.onopen = function(e){
    console.log("Open");
};
```

**Reply by hst-m** [reply]
Web browser client? Very basic stuff. Now you need to add auto-reconnect code. Try not to spam repo in unrelated topics, this is very 101 stuff. This topic is "Installation on CentOS 8" where you failed to do basic install
